UI Page Layout

nhaagen opened the pull request.

Commit:
Basic document structur created

Commit:
Added mockery and unit test definition

Commit:
Installation instructions added

Commit:
phpunit cli commands added

Commit:
Naming concepts enhanced and test structure described

Commit:
FIRST acronym added

Commit:
First part of the correct principle added

corro committed:
Fix markdown errors

Commit:
Merge branch 'feature/5-3/UnitTestInvestigationFeedback' into 'feature/5-3/UnitTestInvestigation'

Fix markdown errors

See merge request !12

Commit:
anti example section added

Commit:
Merge branch 'feature/5-3/UnitTestInvestigation' of git.studer-raimann.ch:ILIAS-eLearning/ILIAS into feature/5-3/UnitTestInvestigation

Commit:
SOLID principles added

Commit:
Added chapter how to use the DIC and example test regretion

Added chapter how use the DIC. In addition an other anti example has been added to show the test regression problem in ILIAS. Finally some small adjustments were made.

Commit:
Added alias example

Commit:
Added normal example unit test example

Commit:
Added a example to mock parent classes

Commit:
Added fluent interface example

Commit:
Added FAQ and How to run tests in PHPStorm

Commit:
Indentations changed from space to tabs in order to fix example display

Commit:
Two missing tabs added

Commit:
Composer setup updated as well as xcode setup

Commit:
Removed unneeded xdebug configuration

Commit:
First part of the discused changes

Commit:
Collectors and test suite configuration added

Commit:
Mockery overload explained

Commit:
Added foreword

Commit:
Added naming proposal

Commit:
Removed unneeded spaces on the left table side

Commit:
Added fun fact for plane landing test

chfsx committed:
Inserted TOC

chfsx committed:
Merge branch 'trunk' into feature/5-3/UnitTestInvestigation

* trunk: (395 commits)
  Fixed permission settings
  fixed bug #21550: Metric question of subtype interval breaks pdf export of the survey
  [Fix 21516]Cal app: Files ordered alphabetically in modals and agenda.
  [Fix 21568] Svy: results/details tab
  0021305: Statistics of groups is not available - What is the cause?
  fixed bug #21461: Failed test: Ankerlinks einfügen
  fixed bug #21528: Shorter and more streamlined label
  fixed bug #21569: sub-tab Participant View is not show, but error message comes up
  fixed bug #21574: cont_import error: cont_transl_master_language_not_allowed
  fixed bug #21543: Failed Test: Filter has no default value
  Added string cast.
  0021549: Export of Excelsheet for Tabel
  fixed bug 21310: Failed test: Mehrere SRT-Dateien hochladen
  fixed mixed content issue, #19506: Mediacast cannot play Youtube Videos
  0020730: Apache Authentication: Missing feature in 5.2.x -> Autom. process Apache SSO
  [Fix 21506] Calendar: language vars and improvements.
  One line too much…
  Added missing language variables for confirmation message #iass_amend_saved# in English and German
  Translation and label improvement
  Another improvement in the lang var
  ...

Commit:
Add quotation and spelling / small rewording

Commit:
Line breaks at roughly 80 chars

Commit:
Add note that the after step should hardly ever be required

Commit:
Removed installation guide of xdebug for macOS and Windows

Commit:
Add explenation for book choice

Commit:
Link external documentation with visible url

Commit:
Remove old documentation link

Commit:
Changed quotation of e-books from location to chapter

Commit:
Add source for unit-test definition

Commit:
Add source for php unit description

Commit:
Improved ILIAS installation bound chapter titles

Commit:
Add namespace to test naming

Commit:
Add chapter which describes the ILIAS installation bound tests

Commit:
Add chapter about how to get installation specific information for unit testing configuration.

Commit:
Fix tripple A for button factory test

Commit:
Shorten button factory tests

Commit:
Tighten unit-test structure chapter

mjansenDatabay committed:
User: Fixed #22440

mjansenDatabay committed:
User: Fixed #22440 (ilContext::doAuthentication() does not return TRUE for the web context anymore)

mjansenDatabay committed:
User: Fixed #22440 (baseClass was missing when building redirect URL)

Commit:
Improved ButtonTest code example

Commit:
Changed example solution method names to lower camel case.

Commit:
Change the proposed solution of the ButtonTest example.

chfsx committed:
supported versions 5.3

nhaagen committed:
glyph-renderer and iconographic button

nhaagen committed:
PageLayoutRevision: Sidebar, Metabar and Page with their respective components and tools

gvollbach committed:
fixed #22504, scaling of page

smeyer-ilias committed:
0022424: ilObjectPermissionStatusGUI: [] operator not supported for strings

klees commented on src/UI/Implementation/Render/FSLoader.php:
```diff
@@ -13,7 +13,7 @@
  * * create a new implementation of RendererFactory in the implementation folder of that
  *   component
- * * introduce it as a dependency of this loader
+ * * introduce it as a dependency of this loader (and load at ilInitialisation::initUIFramework)
```
👍

klees commented on src/UI/Component/Button/Iconographic.php:
```diff
@@ -0,0 +1,26 @@
+	 * @param 	bool 	$state
+	 * @return 	Iconographic
+	 */
+	public function withEngagedState($state);
```
Ich glaube hier werden wir den Namen nochmal ändern, aber nicht jetzt: https://www.ilias.de/mantis/view.php?id=21451

klees commented on src/UI/Component/Button/Factory.php:
```diff
@@ -209,5 +209,46 @@ public function month($default);
+	 *     Both, icon and text, form an entity an can only be operated together.
+	 *     The button MAY have a "down"-state: When the button is used to toggle
+	 *     the visibility of a slate, it stays engaged until the slate is closed
+	 *     or it is clicked again.
```
Hier werde ich einmal drüberaxten.

klees commented on src/UI/Component/Button/Iconographic.php:
```diff
@@ -0,0 +1,26 @@
+	/**
+	 * @return ILIAS\UI\Component\Icon\Icon | \ILIAS\UI\Component\Glyph\Glyph
+	 */
+	public function getIcon();
```
Gibt es irgendeinen guten Überbegriff für Icon und Glyph? Sonst finde ich den Namen verwirrend und würde "getIconOrGlyph" vorziehen. Ebenso im Konstructor in der Factory.

nhaagen replied:
getGraphical? getGraphicalElement?
getIconOrGlyph ist vermutlich richtig.

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,187 @@
+     *     should permamently be in sight of the user.
+     *     Contents of the bar are never modified except by administrative configuration.
+     *     The metabar is not for navigation; user-reactions to system notifications
+     *     are OK, though.
```
Den letzten Satz verstehe ich nicht.

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,187 @@
+     *     further elements are placed on the right hand side.
+     *
+     *   effect: >
+     *     The bar is rendered horizontally on top of the page.
```
Auch über diesen Abschnitt werde ich drüberaxten.

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,187 @@
+     *   usage:
+     *     1: The metabar is unique for the page - there MUST be but one.
+     *     2: Elements in the metabar MUST NOT vary according to context.
+     *     3: Metabar MUST NOT have exclusively navigational elements.
```
Bitte erklären, ich formuliere dann ggf. neu.

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,187 @@
+     *
+     *   composition:
+     *     1: The bar MUST contain the logo.
+     *     2: The bar SHOULD contain prompts.
```
Und auch einen Logout-Button, oder?

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,187 @@
+     *
+     * @return  \ILIAS\UI\Component\Layout\Metabar
+     */
+    public function metabar();
```
Warum erzwingen wir hier nicht bereits, dass prompts eingepflegt werden?

klees replied:
Warum heißt das eigentlich nicht topbar?

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,187 @@
+     *     like the user's profile or administrative settings.
+     *
+     *     The contents of the bar are never modified by changing context,
+     *     but may vary according to e.g. the current user's permissions.
```
...or the settings of the installation.

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,187 @@
+     *
+     *     In a desktop environment, a vertical bar is rendered on the left side
+     *     of the screen covering the full height minus the header-area.
+     *     Entries are aligned vertically.
```
Gehört imo zu `composition`.

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,187 @@
+     *     The contents of the bar itself will not scroll.
+     *
+     *     Width of content- and footer-area is limited to a maximum of the
+     *     overall available width minus that of the bar.
```
Ich glaube das stimmt nicht, oder ich verstehe es nicht.

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,187 @@
+     *     For mobile devices, the bar is rendered horizontally on the bottom
+     *     of the screen with the entries aligned horizontally.
+     *     Again, entries will become inacessible, if the window/screen is smaller
+     *     than the width of all entries summed up.
```
Den Absatz würde ich weglassen. Das ist imo noch nicht final geklärt.

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,187 @@
+     *     There is but one active slate in the bar.
+     *     Iconographic buttons in the sidebar are stateful, i.e. they have a
+     *     pressed-status that can either be toggled by clicking the same button again
+     *     or by clicking a different button.
```
Auch hier würde ich nochmal drüberaxten, sobald wir die Fragen geklärt haben.

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,187 @@
+     *     Tab Bar: >
+     *       The sidebar (and its components) shall not be used to substitute
+     *       functionality available at objects, such as settings, members or
+     *       learning progress. Those remain in the Tab Bar.
```
👍

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,187 @@
+     *     Content Actions: >
+     *       Also, adding new items, the actions-menu (with comments, notes and tags),
+     *       moving, linking or deleting objects and the like are not part of
+     *       the sidebar.
```
👍

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,187 @@
+     *       in the Notification Center.
+     *       The direction of communication for the sidebar is "user to system",
+     *       while the direction is "system to user" in the Notification Center.
+     *       However, navigation from both components can lead to the same page.
```
👍

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,187 @@
+     *     Modal: >
+     *       Forms with the intention of modifying the content are placed in modals
+     *       or on the content-page.
+     *
```
Sollte hier nicht vielleicht auch die Slate als Rivale stehen? IMO ist der Unterschied, dass die Slate auch komplexere Navigation enthalten kann und die Sidebar die Hauptnavigation ist.

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,187 @@
+     *
+     * rules:
+     *   usage:
+     *     1: The sidebar is unique for the page - there MUST be but one.
```
Vielleicht zu SHOULD verkleinern, für KIOSK-Mode, etc.

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,187 @@
+     *     1: The sidebar is unique for the page - there MUST be but one.
+     *
+     *   composition:
+     *     1: The bar MUST NOT contain items other than buttons.
```
"_iconographic_ buttons", oder?

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,187 @@
+     *
+     *   composition:
+     *     1: The bar MUST NOT contain items other than buttons.
+     *     2: The bar MUST contain at least one button.
```
s.o.?

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,187 @@
+     *
+     *   style:
+     *     1: The bar MUST have a fixed witdth (desktop).
+     *     2: The bar MUST have a fixed height (mobile).
```
Mobile weg.

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,187 @@
+     *        navigational options within the bar (open a slate)
+     *        OR actually invoke navigation, i.e. change the location/content
+     *        of the current page.
+     *     2: Elements in the bar MUST NOT open a modal or window.
```
👍 !!

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,187 @@
+     * ----
+     *
+     * @param  \ILIAS\UI\Component\Layout\SidebarEntry[]    $entries
+     * @param  int|null  $active
```
IMO müsste das mixed sein, weil sich das auf die Indizes aus `$entries` bezieht, oder? Das müsste dann hier auch kommentiert werden.

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,187 @@
+     *   purpose: >
+     *     The sidebar entry bundles a button and a slate.
+     *   composition: >
+     *     There is no composition of this component.
```
Hier imo stattdessen reinschreiben, dass das selbst keine visuelle Ausprägung hat, sonder in Button und Slate zerfällt.

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,187 @@
+     * @param  int|null  $active
+     * @return  \ILIAS\UI\Component\Layout\Sidebar
+     */
+    public function sidebar($entries, $active=null);
```
Auf `array $entries` hinten.

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,187 @@
+     *
+     * ----
+     *
+     * @param  \ILIAS\UI\Component\Button\Iconographic | \ILIAS\UI\Component\Glyph\Glyph   $button
```
`Glyph` ist imo nicht mehr nötig, das haben wir ja schon im Iconographic Button. Dann kann unten auch ein entsprechender Hint eingebaut werden.

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,187 @@
+      *     the page thus is the user's view upon ILIAS in total.
+	 *
+      *   composition: >
+      *      2do.
```
Will do.

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,187 @@
+      *   usage:
+      *     1: The page MUST be rendered with content.
+      *     1: The page SHOULD be rendered with a Metabar.
+      *     2: The page SHOULD be rendered with a Sidebar.
```
S.o., Widersprich zu Sidebar-Regeln.

klees commented on src/UI/Component/Layout/Metabar.php:
```diff
@@ -0,0 +1,34 @@
+	 * @param 	\ILIAS\UI\Component\Image\Image
+	 * @return 	Metabar
+	 */
+	public function withLogo($logo);
```
Hint einbauen und oben Parameternamen ergänzen.

klees commented on src/UI/Component/Layout/Metabar.php:
```diff
@@ -0,0 +1,34 @@
+	 * @param 	mixed 	$element
+	 * @return 	Metabar
+	 */
+	public function withElement($element);
```
Wir haben hier keine gemeinsame Vaterklasse für Dinge, die da rein dürfen? Warum beschränken wir nicht direkt auf die drei bekannten? Das könnten wir dann sogar schon im Konstruktor machen, ebenso das Image verlangen.

klees commented on src/UI/Component/Layout/Page.php:
```diff
@@ -0,0 +1,44 @@
+	public function withMetabar(C\Layout\Metabar $metabar);
+
+	/**
+	 * @return 	ILIAS\UI\Component\Layout\Metabar
```
Könnte auch null sein.

klees commented on src/UI/Component/Layout/Page.php:
```diff
@@ -0,0 +1,44 @@
+	 * @param 	ILIAS\UI\Component\Layout\Metabar $metabar
+	 * @return 	Page
+	 */
+	public function withMetabar(C\Layout\Metabar $metabar);
```
Vielleicht die `Metabar` im Konstruktor verlangen, um die SHOULD-Regel zu unterstreichen und dann hier `withoutMetabar`?

klees commented on src/UI/Component/Layout/Page.php:
```diff
@@ -0,0 +1,44 @@
+	public function withSidebar(C\Layout\Sidebar $sidebar);
+
+	/**
+	 * @return 	ILIAS\UI\Component\Layout\Sidebar
```
Könnte auch null sein.

klees commented on src/UI/Component/Layout/Page.php:
```diff
@@ -0,0 +1,44 @@
+	 * @param 	ILIAS\UI\Component\Layout\Sidebar 	$sidebar
+	 * @return 	Page
+	 */
+	public function withSidebar(C\Layout\Sidebar $sidebar);
```
Hier gilt das gleiche wie bei der Metabar?

klees commented on src/UI/Component/Layout/Sidebar.php:
```diff
@@ -0,0 +1,41 @@
+	/**
+	 * @return 	Signal
+	 */
+	public function getEntryClickSignal();
```
Hier sieht man wieder, dass die Signals unterspezifiert sind, oder? Ich habe mich gerade gefragt, woher ich den weiß, welcher Entry geklickt werden soll. Weiß ich aber hier garnicht, wird per Option (glorified array) an das Signal drangebamselt, oder?

klees commented on src/UI/Component/Layout/Sidebar.php:
```diff
@@ -0,0 +1,41 @@
+
+	/**
+	 * This is the index of the active entry.
+	 * @return 	int
```
Ähnliche Frage wie bei der Factory oben. Ist das nicht vielleicht mixed?

klees commented on src/UI/Component/Layout/Sidebar.php:
```diff
@@ -0,0 +1,41 @@
+	 * @param 	int 	$active
+	 * @return 	Sidebar
+	 */
+	public function withActive($active);
```
Wird das aktuell irgendwo verwendet?

klees commented on src/UI/Component/MainControls/Factory.php:
```diff
@@ -0,0 +1,36 @@
+	 *   purpose: >
+	 *     Prompts are notifications from the system to the user.
+	 *
+	 * ----
```
Hier werde ich drüberaxten.

klees commented on src/UI/Component/MainControls/Factory.php:
```diff
@@ -0,0 +1,36 @@
+	 *     Menu elements are wrappers for other components. They are used to
+	 *     (sub-)structure navigational elements.
+	 *     Their functionality is of a mere visual nature, i.e. displaying or
+	 *     hiding underlying elements.
```
Hier werde ich eventuell nochmal drüber gehen.

klees commented on src/UI/Component/MainControls/Menu/Factory.php:
```diff
@@ -0,0 +1,76 @@
+	 *     The slate will allways have a "close"-button at its bottom.
+	 *
+	 *     When content-length exceeds the slate's height, the area above the
+	 *     close button will start scrolling vertically with a scrollbar on the right.
```
Hier werde ich nochmal drüberaxten.

klees commented on src/UI/Component/MainControls/Menu/Factory.php:
```diff
@@ -0,0 +1,76 @@
+	 *     Planks MAY contain further planks as well as other components.
+	 *
+	 *   effect: >
+	 *     A plank can be expanded/collapsed by operating the glyph before the title.
```
Gibt's das so noch?

klees commented on src/UI/Component/MainControls/Menu/Factory.php:
```diff
@@ -0,0 +1,76 @@
+	 *     Dropdown: >
+	 *       The dropdown renders above underlying content, while the plank will
+	 *       push down other elements when expanded.
+	 *       Planks can be cascaded.
```
Ich glaube der ganze Absatz ist nicht aktuell.

klees commented on src/UI/Component/MainControls/Menu/Plank.php:
```diff
@@ -0,0 +1,34 @@
+	/**
+	 * @return 	Signal
+	 */
+	//public function getToggleSignal();
```
?

klees commented on src/UI/Component/MainControls/Menu/Slate.php:
```diff
@@ -0,0 +1,69 @@
+	 * @param 	Plank[] 	$planks
+	 * @return 	Slate
+	 */
+	public function withPlanks(array $planks);
```
Wollen wir hier vllt. noch "withAdditionalPlank"?

klees commented on src/UI/Component/MainControls/Prompts/Factory.php:
```diff
@@ -0,0 +1,84 @@
+     *
+     * rules:
+     *   usage:
+     *     1: There MUST be but one notification center on the page.
```
Muss nicht vielleicht das Notification-Center in der Metabar sein? Oder sollte?

klees commented on src/UI/Component/MainControls/Prompts/Factory.php:
```diff
@@ -0,0 +1,84 @@
+     *     Notification Center: >
+     *        The Notification Center is for permament notes, such as mails in
+     *        the inbox. Also, the contents are limited to more counter-glyphs
+     *        which will change the context when operated.
```
Hier axte ich nochmal drüber. Das Awareness-Tool ist eigentlich schlecht benannt und bündelt Funktionen über Benutzer ("Who is online", Kontakte) und nicht über allgemeine Systemevents.

klees commented on src/UI/Component/MainControls/Prompts/Factory.php:
```diff
@@ -0,0 +1,84 @@
+     *
+     * rules:
+     *   usage:
+     *     1: There MUST be but one awareness-tool on the page.
```
Wie beim Notification Center sollte sich das imo auf die Metabar beziehen.

klees reviewed (commented):
Hallo Nils!

Sehr sehr gute Arbeit. Bin noch nicht bis zum Ende gekommen, habe aber schon einiges an Anmerkungen eingepflegt. Vieles müssen wir wahrscheinlich besprechen, einiges werde ich übernehmen. Den Rest werde ich morgen durchsehen. Es wäre aber super, wenn du mit meinen bestehenden Kommentaren morgen vormittag schonmal beginnen könntest (so nichts vom Release dringender ist).

Beste Grüße!

nhaagen committed:
adjust misleading function/var-name

nhaagen committed:
update docs and factories

nhaagen committed:
remove inline page-example

klees commented on src/UI/Factory.php:
```diff
@@ -707,4 +707,42 @@ public function dropzone();
+	 *       Layout components are page-elements used for the basic construction of
+	 *       the user interface. In ultimo, also the page itself is included here.
+	 *
+	 * ---
```
Hier werde ich nochmal drüberaxten.

klees commented on src/UI/Factory.php:
```diff
@@ -707,4 +707,42 @@ public function dropzone();
+	 *   usage:
+	 *     1: MainControls MUST NOT change persistent data in any way.
+	 *   accessibility:
+	 *     1:
```
Hier werde ich nochmal drüberaxten.

klees commented on src/UI/Implementation/Component/Glyph/GlyphRendererFactory.php:
```diff
@@ -0,0 +1,20 @@
+		}
+		if( in_array('NotificationCenterPromptsMainControls', $contexts)) {
+			return new NotificationCenterContextRenderer($this->ui_factory, $this->tpl_factory, $this->lng, $this->js_binding);
+		}
```
Warum brauchen wir zwei neue Kontexte?

nhaagen replied:
Im Button brauche ich einen <span> und im Notification Center noch das label.

klees commented on src/UI/Implementation/Component/Layout/Renderer.php:
```diff
@@ -0,0 +1,196 @@
+        $tpl = $this->getTemplate("tpl.sidebar.html", true, true);
+        $entry_signal = $component->getEntryClickSignal();
+        $counter = 0;
+        foreach ($component->getEntries() as $entry) {
```
Ich würde vorschlagen, den $counter nicht zu nutzen sondern stattdessen `$entry` in `$index => $entry` zu verwandeln und dann $index zu nutzen. Das scheint mir robuster.

klees commented on src/UI/Implementation/Component/Layout/Renderer.php:
```diff
@@ -0,0 +1,196 @@
+                } else {
+                    triggerer.addClass(down_class);
+                    triggerer.attr('aria-pressed', true);
+                }
```
Das vielleicht in eine Javascript-Datei auslagern?

klees commented on src/UI/Implementation/Component/Layout/Renderer.php:
```diff
@@ -0,0 +1,196 @@
+            $tpl->setVariable("ELEMENT", $default_renderer->render($element));
+            $tpl->parseCurrentBlock();
+        }
+        //2do: this must be a glyph with proper action
```
Do it.

klees commented on src/UI/Implementation/Component/Layout/Renderer.php:
```diff
@@ -0,0 +1,196 @@
+    protected function setHeaderVars($tpl) {
+
+        global $DIC;
+        $il_tpl = $DIC["tpl"];
```
Diese Abhängigkeit sollten wir explizit machen. Letztlich müsste ich doch eigentlich auf der Page die benötigten JS- und CSS-Dateien als Properties angeben können, um die dann hier zu nutzen. Alternativ könnte man irgendein Interface mit `getJSRessources` und `getCSSRessources` oder so (ähnlich JILL?) einführen und dass nutzen. Ich glaub, dass müssen wir nicht unbedingt noch jetzt machen, aber die Überlegung sollten wir zumindestens mal kurz machen und hier dokumentieren.

klees commented on src/UI/Implementation/Component/Layout/Renderer.php:
```diff
@@ -0,0 +1,196 @@
+use ILIAS\UI\Renderer as RendererInterface;
+use ILIAS\UI\Component;
+
+class Renderer extends AbstractComponentRenderer {
```
Einrückung auf Tabs umstellen bitte.

klees commented on src/UI/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,182 @@
+     *     its button will be rendered in the sidebar with a signal triggering
+     *     its slate.
+     *
+     * ----
```
Ich denke hier sollten wir dokumentieren, dass der `sidebarEntry` den Button mit der Slate verknüpft, oder diese Verknüpfung nicht in dem `SidebarEntry` sondern erst kurz vor knapp machen. Letztlich ist es irritierend, das dies nicht funktioniert:

```
$entry = $f->sidebarEntry($button);
$button2 = $entry->getButton();
assert($button === $button2);
```

klees commented on src/UI/Implementation/Component/MainControls/Menu/Renderer.php:
```diff
@@ -0,0 +1,123 @@
+use ILIAS\UI\Renderer as RendererInterface;
+use ILIAS\UI\Component;
+
+class Renderer extends AbstractComponentRenderer {
```
Indentation nach Tabs bitte.

klees commented on src/UI/Implementation/Component/MainControls/Prompts/Renderer.php:
```diff
@@ -0,0 +1,104 @@
+        $contents = $component->getContents();
+
+        $glyph = $f->glyph()->user("#");
+            //->withOnClick($popover->getShowSignal());
```
Weg?

klees commented on src/UI/examples/Button/Iconographic/base.php:
```diff
@@ -0,0 +1,21 @@
+	$renderer = $DIC->ui()->renderer();
+
+	$ico = $f->icon()->standard('someExample', 'Example');
+	$ico = $ico->withAbbreviation('E')->withSize('medium');
```
Direkt hintereinander für Style. Zwischenspeichern nicht nötig.

klees commented on src/UI/examples/Layout/Page/ui.php:
```diff
@@ -0,0 +1,55 @@
+
+    include_once('src/UI/examples/Layout/Metabar/metabar.php');
+    include_once('src/UI/examples/Layout/Page/page.php');
+    ilInitialisation::initILIAS();
```
Vielleicht den Init-Foo noch in eine Extra-Funcktion auslagern und den Rendering-Foo auch. Dann kann man den Rendering-Foo drüberpacken und sieht das (leider nötige) Chaos nicht?

klees commented on src/UI/examples/Layout/Page/z_auxilliary.php:
```diff
@@ -0,0 +1,132 @@
+    $slate = $f->maincontrols()->menu()->slate(array()); //remove constructor-planks
+    $replace_signal = $slate->getReplaceContentSignal();
+    $slate = $slate->withPlanks(pagedemo_planks1($f, $replace_signal));
+    $entries[] = $f->layout()->sidebarentry($btn, $slate);
```
Ich glaube, dass sieht ohne temporäre Variablen besser aus, können wir mal zusammen ausprobieren.

klees commented on src/UI/examples/Layout/Page/z_auxilliary.php:
```diff
@@ -0,0 +1,132 @@
+function pagedemoSidebar($f) {
+    $icon = $f->icon()->custom('./src/UI/examples/Layout/Page/icon-sb-cockpit.svg', '');
+    $btn = $f->button()->iconographic($icon->withSize('large'), "Button 1", '#');
+    $slate = $f->maincontrols()->menu()->slate(array()); //remove constructor-planks
```
Warum `remove constructor-planks`?

klees reviewed (commented):
Alter Vadda. Ich werde mir morgen die Examples noch mal weiter im Detail ansehen.

xus committed:
[Fix 21553]IlDateTime set user timezone as default only if the object does not have one.

nhaagen committed:
modify action of button in renderer, not in entry

nhaagen committed:
cosmetics, minor cleanups

shecken committed:
Fix usage of class propertys and cleanup tabs/spaces

shecken committed:
Remove class property allocation to function variabe

shecken committed:
Mark export tab as active

shecken committed:
Create function to create an export tab

mjansenDatabay committed:
Chat/OnScreenChat: Fixed deletion

mjansenDatabay committed:
Chat/OnScreenChat: Fixed deletion

mjansenDatabay committed:
Link URL handling

(cherry picked from commit 08ae035)

Commit:
Exclude client ip in tokens

Commit:
Exclude client ip in tokens

shecken committed:
Revert "Remove class property allocation to function variabe"

This reverts commit c010a4c195ec09fc62597a4a586304a20f3146fc.

shecken committed:
Revert "Fix usage of class propertys and cleanup tabs/spaces"

This reverts commit b1017f89ec6c86c2d55dc331adeb1cb5cd7587f7.

nhaagen committed:
added logout-glyph

nhaagen committed:
grey bg for sidebar-buttons

mjansenDatabay committed:
User: Fixed #22440 after Jour Fixe decision

mjansenDatabay committed:
Documentation: Added some hints regarding MySQL/MariaDB innodb_large_prefix

Amstutz committed:
Update docs-guidelines according JF decision

Amstutz committed:
Markdown clean-up

gvollbach committed:
added byline for phantom js version

chfsx committed:
improved interface docu of ilDBInterface

bogenschlag committed:
Update docs-guidelines.md resolved conflict

bogenschlag committed:
Update ini-file.md

Cleaned up headings, rewrapped the text paragraph.

Since plain text readability is important, I propose to format the tables (I used the *Markdown table formatter* package from the *Sublime Text* repo). Otherwise, it seems to me, there is no hope in heck to make sense of the tables in plain text view. They have to be viewed unwrapped, of course. But wrapping the tables IMHO makes no sense either way (and I’m not even sure that wouldn’t break parsing).

What do you think?
(Does this need a JF decision?)

tmuellerdatabay committed:
Fixes \n in assFileUploadFileTableGUI

bheyser committed:
SingleChoice: Standard Answer SingleChoice

matthiaskunkel committed:
Fixed Mantis22519: ilias_de.lang wrong parameter count

alex40724 committed:
support 7.1 with 5.2

Commit:
fix mantis #22534

(cherry picked from commit e5a9f4a)

Amstutz committed:
Merge pull request #873 from mjansenDatabay/docu_mariadb_innodb_large_index

Documentation: Added some hints regarding MySQL/MariaDB innodb_large_…

bheyser committed:
SingleChoice: fixed missing end div for SingleChoice

bheyser committed:
all types: FeedbackRight FeedbackWrong in Preview and Player

mjansenDatabay committed:
Chat: Disabled OnScreen-Chat if chat is disabled in general (bugfix for #22539)

bogenschlag committed:
Update install.md

Apparently I had overlooked one paragraph that should have been wrapped and "needed" it’s unnecessary inline tripple backticks reduced to single ones.

bheyser committed:
all types: FeedbackWrong FeedbackRight for solution/result screens

chfsx committed:
missing changes from release_5-3

bheyser committed:
all types: Submit

bogenschlag committed:
Update contributing.md

Rewrapping, clean-up, added TOC.

bheyser committed:
SingleChoice, MultipleChoice, KprimChoice: completed remaining css classes

bheyser committed:
Kprim Choice: completed templates and purejs data in LM

bheyser committed:
Kprim Choice: removed console.log in LM

bheyser committed:
ErrorText: complete set of relevant css classes

bheyser committed:
ImageMap: compete set of ilc classes

bheyser committed:
ClozeTest: Complete Style Set

bheyser committed:
Numeric: complete set

bheyser committed:
FormulaQuestion: all css classes

bheyser committed:
TextSubset: all classes

bheyser committed:
OrderingQuestion: complete css set

bheyser committed:
OrderingQuestion: QuestionImage

bheyser committed:
OrderingHorizontal: all ilc classes

bheyser committed:
MatchingQuestion: complete ilc classes set

bheyser committed:
TextQuestion: all classes

bheyser committed:
FileUpload: complete ilc css set

bheyser committed:
LongMenu: complete set of ilc css classes

bheyser committed:
Merge branch '54questionStyles' into trunk

alex40724 committed:
news: some code cleaning

chfsx committed:
[FIX] 0020343: static access

Uwe-Kohnle committed:
small improvement for disply in iframe

alex40724 committed:
Update README.md #881

Commit:
Merge pull request #870 from shecken/trunk_export_plugin

Easier way to get the export tab and functionality for Plugins

alex40724 committed:
fixed  Setup: Switching SETTINGS DATA TYPE from TEXT to CLOB does not work (in the Tools area)

alex40724 committed:
fixed bug #22566: Wrong passed status with team upload

smeyer-ilias committed:
Fixed default object settings of print view.

mjansenDatabay committed:
Another fix for links URLs

(cherry picked from commit 4333874)

mjansenDatabay committed:
Another fix for links URLs

(cherry picked from commit e00836a)

mjansenDatabay committed:
Another fix for links URLs

mjansenDatabay committed:
Another fix for links URLs

(cherry picked from commit 98cbb54)

Elwe-Thingol committed:
fix for bug 0022565

Commit:
Use repository selector in data collection

mjansenDatabay committed:
Main Menu History: Workaround for #22588

mjansenDatabay committed:
Main Menu History: Workaround for #22588

matthiaskunkel committed:
Fixed Mantis 22553: Misleading descriptions

mjansenDatabay committed:
Footer: Fixed system contacts link

chfsx committed:
Merge remote-tracking branch 'gitlab/fix/0022336/exclude_client_ip_in_tokens' into trunk

chfsx committed:
updated TOC

chfsx committed:
we're not there yet!

chfsx committed:
FIRST

bheyser committed:
fixed mantis 0022604: Typo seid/seit

bheyser committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into trunk

chfsx committed:
fixed WAC tests

Commit:
reload ctrl structure for fix 22524

matthiaskunkel committed:
Some improvements in German lang file

matthiaskunkel committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into trunk

klees committed:
warning regarding mockery

smeyer-ilias committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into trunk

smeyer-ilias committed:
Fixed 0022312: CONSTRAINT-Violations during Role editing.

smeyer-ilias committed:
Bugfix for duplicate entry error

Commit:
Use repository selector in data collection (Fix async)

mjansenDatabay committed:
Certificate: Fixed #22616

bheyser committed:
removed additional line at end of file by removing the php end marker

chfsx committed:
removed non used node_module

alex40724 committed:
fixed bug #22612: ilCtrlException when clicking on View in repository root

Commit:
Use repository selector in data collection (Fix ilPropertyFormGUI sub class)

Commit:
Use repository selector in data collection

klees committed:
UI/Button: allow buttons to change the view's status as well.

klees committed:
UI/Button: clearly state: no navigation

klees committed:
UI/Button: added description what a view change might be.

Elwe-Thingol committed:
fixed the problem with to many results

Uwe-Kohnle committed:
SOP: more time to get content

smeyer-ilias committed:
0022625: \ilECSMappingUtils::getCourseValueByMappingAttribute returns string

xus committed:
[Fix 22526]Survey show question titles setting

xus committed:
[Fix 22638] Custom Metadata in list view: strip_tags only non location type elements.

matthiaskunkel committed:
Fixed Mantis #22640, thanks to Chris Potter

matthiaskunkel committed:
Fixed Msntis #22639

bheyser committed:
added labels for mc/sc answer options in lm template and completed answertext class for options in assessment templates

bheyser committed:
fixed type in the wrong window accident

Commit:
Merge pull request #888 from studer-raimann/fix/22524/use_repository_selector_in_data_collection

Use repository selector in data collection

Commit:
Remove unkown expression for old nodejs versions

Amstutz committed:
Merge pull request #897 from klees/trunk_ui_button

UI/Button: allow buttons to change the view's status

bheyser committed:
fixed/improved current situation for question content reused in learning module (p tags, latex, feedback, type specific content fields)

bheyser committed:
removed error files added by mistake

bheyser committed:
fixed re-save unqualified cloze gap objects due to heavy overload of setClozeText() setter // removed save call not neccessary

mjansenDatabay committed:
Merge pull request #906 from legionth/fix-start-chatserver-trunk

[trunk] Remove unkown expression for old nodejs versions

alex40724 committed:
added check for step 5242, see #22649: Error bei DB-Update Step <0005242> - wenn File nicht verschoben / gelöscht werden kann

alex40724 committed:
added logging

Amstutz committed:
UI Tests: Corrected ugly Warning: Declaration of ilIndependentTemplate::get($part = NULL) should be compatible with ... warning

Commit:
Remove unused  class

mjansenDatabay committed:
Merge pull request #912 from legionth/remove-unneeded-cron-class

Remove unused cron classes

chfsx committed:
[FIX] 12860: Filtering mb4-strings

kyro46 committed:
Add log entry if solution is willingly deleted

bheyser committed:
fixed static call to non static instance method

bheyser committed:
added missing check for something

alex40724 committed:
use central wrapper

smeyer-ilias committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into trunk

smeyer-ilias committed:
\ilObjectCopyGUI::copyMultipleNonContainer: Duplicate code block

mjansenDatabay committed:
Chat Administration: Used central services

mjansenDatabay committed:
User: Some changes regarding system user handling

mjansenDatabay committed:
Workflow Engine: Switched to centralized approach

(cherry picked from commit 09d5826)

mjansenDatabay committed:
User: Some changes regarding system user handling

mjansenDatabay committed:
Merge remote-tracking branch 'mjansen/changed-handling-of-root' into trunk

matthiaskunkel committed:
Fixed Mantis #22618: Streamling Course and Group Settings

smeyer-ilias committed:
Merge pull request #860 from mjansenDatabay/bugfix-22440

Bugfix 22440

alex40724 committed:
ext list

alex40724 committed:
use wrappers

mjansenDatabay committed:
Course/Group: Fixed #123082

(cherry picked from commit c89e37b)

mjansenDatabay committed:
Cron: Added check

alex40724 committed:
abandonded old import version

alex40724 committed:
removed unused code

mjansenDatabay committed:
Session Reminder/DIC: #22631

(cherry picked from commit 43cb475)

alex40724 committed:
added woff extension

alex40724 committed:
fixed bug #22675: User creation: Upload error: File not found

mjansenDatabay committed:
Chat Administration: Form handling ...

mjansenDatabay committed:
Added bpmn extension

mjansenDatabay committed:
Workflow Engine: Some process definition handling changes after form submit

chfsx committed:
WIP

Amstutz committed:
Merge pull request #903 from ILIAS-TechnicalBoard/feature/5-3/UnitTestInvestigation

Unit Test Readme

chfsx committed:
[FIX] 0022682 \ilDBPdoMySQL: Possible performance issue

alex40724 committed:
central rename using white list

mjansenDatabay committed:
Mail/Forum: Used central wrapper

Elwe-Thingol committed:
added g flag to regex in order to replace all occurences of mb4 strings

chfsx committed:
Merge pull request #915 from studer-raimann/fix/5-3/regex-in-santizemb4string-does-not-replace-all-mb4-strings-bug

added g flag to regex in order to replace all occurences of mb4 strings

chfsx committed:
Revert "added g flag to regex in order to replace all occurences of mb4 strings"

This reverts commit 70064ef239f0623c5c386d67843190d710e9e650.

chfsx committed:
Merge branch 'trunk' into fix/5-4/file-security-renaming

* trunk: (63 commits)
  Revert "added g flag to regex in order to replace all occurences of mb4 strings"
  added g flag to regex in order to replace all occurences of mb4 strings
  Mail/Forum: Used central wrapper
  central rename using white list
  [FIX] 0022682 \ilDBPdoMySQL: Possible performance issue
  Workflow Engine: Some process definition handling changes after form submit
  Added bpmn extension
  Chat Administration: Form handling ...
  fixed bug #22675: User creation: Upload error: File not found
  added woff extension
  Session Reminder/DIC: #22631
  warning regarding mockery
  FIRST
  we're not there yet!
  updated TOC
  Change the proposed solution of the ButtonTest example.
  Changed example solution method names to lower camel case.
  Improved ButtonTest code example
  Tighten unit-test structure chapter
  Shorten button factory tests
  ...

alex40724 committed:
fixed dir renaming; added extensions needed by scorm

chfsx committed:
implemented whitelist in FileUploadService

smeyer-ilias committed:
Added logging.

smeyer-ilias committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into trunk

smeyer-ilias committed:
removed debug

chfsx committed:
Merge branch 'fix/5-4/file-upload-handling' into trunk

alex40724 committed:
scorm extensions

alex40724 committed:
scorm extensions

Commit:
switched to ilFileUtils

Uwe-Kohnle committed:
add ilUtil to rename for SCORM

nhaagen committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into UI_PageLayout

catkrahl committed:
merge

nhaagen committed:
fixed example
